### PR TITLE
Support being embedded within a standalone executable

### PR DESCRIPTION
**Pull request by ofek**
The goal of this is to support [PyOxidizer](https://github.com/indygreg/PyOxidizer) so virtualenv, and projects that depend on it like tox or Hatch, can be distributed as a standalone executable.

The only change is the continued functionally when there is no `__file__` global such as when resources are loaded from memory. More info: https://pyoxidizer.readthedocs.io/en/latest/oxidized_importer_resource_files.html

You can build by running `tox -e oxidize`.

The only way this currently works is:

```
❯ .\app\build\x86_64-pc-windows-msvc\debug\exe\virtualenv.exe testvenv --with-traceback -p C:\Users\ofek\AppData\Local\Programs\Python\Python39\python.exe --seeder pip --system-site-packages
created virtual environment CPython3.9.4.final.0-64 in 3198ms
  creator CPython3Windows(dest=C:\Users\ofek\Desktop\code\virtualenv\testvenv, clear=False, no_vcs_ignore=False, global=True)
  seeder PipInvoke(download=False, pip=bundle, setuptools=bundle, wheel=bundle)
    added seed packages: pip==22.1.2, setuptools==62.6.0, wheel==0.37.1
  activators BashActivator,BatchActivator,FishActivator,NushellActivator,PowerShellActivator,PythonActivator
```

Everything else has a failure scenario, which we can iterate on.

## Built-in Python

```
❯ .\app\build\x86_64-pc-windows-msvc\debug\exe\virtualenv.exe testvenv --with-traceback
Traceback (most recent call last):
  File "runpy", line 196, in _run_module_as_main
  File "runpy", line 86, in _run_code
  File "virtualenv.__main__", line 80, in <module>
  File "virtualenv.__main__", line 65, in run_with_catch
  File "virtualenv.__main__", line 18, in run
  File "virtualenv.run", line 32, in cli_run
  File "virtualenv.run.session", line 46, in run
  File "virtualenv.run.session", line 53, in _create
  File "virtualenv.create.creator", line 177, in run
  File "virtualenv.create.via_global_ref.builtin.via_global_self_do", line 97, in create
  File "virtualenv.create.via_global_ref.builtin.ref", line 157, in run
  File "virtualenv.util.path._sync", line 52, in copy
  File "shutil", line 417, in copy
  File "shutil", line 254, in copyfile
FileNotFoundError: [Errno 2] No such file or directory: 'C:\\Users\\ofek\\Desktop\\code\\virtualenv\\APP\\BUILD\\X86_64-PC-WINDOWS-MSVC\\DEBUG\\EXE\\pythonw.exe'
```

@indygreg is there any way to use `sys.executable`?

## Seeder app data

```
❯ .\app\build\x86_64-pc-windows-msvc\debug\exe\virtualenv.exe testvenv --with-traceback -p C:\Users\ofek\AppData\Local\Programs\Python\Python39\python.exe
Traceback (most recent call last):
  File "runpy", line 196, in _run_module_as_main
  File "runpy", line 86, in _run_code
  File "virtualenv.__main__", line 80, in <module>
  File "virtualenv.__main__", line 65, in run_with_catch
  File "virtualenv.__main__", line 18, in run
  File "virtualenv.run", line 32, in cli_run
  File "virtualenv.run.session", line 47, in run
  File "virtualenv.run.session", line 60, in _seed
  File "virtualenv.seed.embed.via_app_data.via_app_data", line 72, in run
RuntimeError: failed to build image wheel, pip because:
Traceback (most recent call last):
  File "virtualenv.seed.embed.via_app_data.via_app_data", line 58, in _install
    installer.install(creator.interpreter.version_info)
  File "virtualenv.seed.embed.via_app_data.pip_install.base", line 44, in install
    consoles.update(self._create_console_entry_point(name, module, script_dir, version_info))
  File "virtualenv.seed.embed.via_app_data.pip_install.base", line 147, in _create_console_entry_point
    new_files = maker.make(specification)
  File "distlib.scripts", line 417, in make
    self._make_script(entry, filenames, options=options)
  File "distlib.scripts", line 317, in _make_script
    self._write_script(scriptnames, shebang, script, filenames, ext)
  File "virtualenv.seed.embed.via_app_data.pip_install.base", line 201, in _write_script
    super(ScriptMakerCustom, self)._write_script(names, shebang, script_bytes, filenames, ext)
  File "distlib.scripts", line 247, in _write_script
    launcher = self._get_launcher('t')
  File "distlib.scripts", line 392, in _get_launcher
    resource = finder(distlib_package).find(name)
  File "distlib.resources", line 332, in finder
    raise DistlibException('Unable to locate finder for %r' % package)
distlib.DistlibException: Unable to locate finder for 'distlib'
```

No idea

## Seeder pip

```
❯ .\app\build\x86_64-pc-windows-msvc\debug\exe\virtualenv.exe testvenv --with-traceback -p C:\Users\ofek\AppData\Local\Programs\Python\Python39\python.exe --seeder pip
C:\Users\ofek\Desktop\code\virtualenv\testvenv\Scripts\python.exe: No module named pip
Traceback (most recent call last):
  File "runpy", line 196, in _run_module_as_main
  File "runpy", line 86, in _run_code
  File "virtualenv.__main__", line 80, in <module>
  File "virtualenv.__main__", line 65, in run_with_catch
  File "virtualenv.__main__", line 18, in run
  File "virtualenv.run", line 32, in cli_run
  File "virtualenv.run.session", line 47, in run
  File "virtualenv.run.session", line 60, in _seed
  File "virtualenv.seed.embed.pip_invoke", line 28, in run
  File "virtualenv.seed.embed.pip_invoke", line 36, in _execute
RuntimeError: failed seed with code 1
```

I probably just missed spots here b/c I don't know the codebase.

-----

Interestingly, the `Seeder pip` actually succeeds after a failed `Seeder app data`.

### CODE REVIEW

File: app/pyoxidizer.bzl [hunk 1]
```diff
@@ -0,0 +1,36 @@
+# This file defines how PyOxidizer application building and packaging is
```
Comment (gaborbernat) [inline app/pyoxidizer.bzl:1]: Can you align this to 120 characters please?

File: src/virtualenv/create/via_global_ref/_virtualenv.py [hunk 2]
```diff
@@ -4,7 +4,10 @@
 import os
 import sys
 
-VIRTUALENV_PATCH_FILE = os.path.join(__file__)
+try:
+    VIRTUALENV_PATCH_FILE = os.path.join(__file__)
+except NameError:
+    VIRTUALENV_PATCH_FILE = None
```
Comment (gaborbernat) [inline src/virtualenv/create/via_global_ref/_virtualenv.py:10]: I'm not a big fan of try/catch; can we instead just if/else with __file__ is None?

**Review by gaborbernat — COMMENTED** [review]
For what's worth I'd like to drop Python 2 support for virtualenv just didn't yet have time to add it. Not sure if keeping it Python 3 only makes it easier.

File: app/pyoxidizer.bzl (hunk 1 above)
Comment (ofek) [reply app/pyoxidizer.bzl:1]: Done!

File: src/virtualenv/create/via_global_ref/_virtualenv.py (hunk 2 above)
Comment (ofek) [reply src/virtualenv/create/via_global_ref/_virtualenv.py:10]: Neither am I but the issue is that `__file__` is undefined

**Comment by ofek** [comment]
Not really, it wouldn't change much

**Comment by indygreg** [comment]
> @indygreg is there any way to use `sys.executable`?

Assuming you are talking about an app in an embedded interpreter / its own executable wanting to run a `python` like process, that isn't currently possible without writing a custom Rust shim to perform the custom `argv` dispatch logic.

We can't safely expose `sys.executable` from application binaries because of the limitation documented at https://gregoryszorc.com/docs/pyoxidizer/main/pyoxidizer_technotes.html#sys-executable-is-a-string-instead-of-list.

There is https://gregoryszorc.com/docs/pyoxidizer/main/pyoxidizer_config_type_python_interpreter_config.html#starlark_pyoxidizer.PythonInterpreterConfig.executable to allow you to define `sys.executable` to a static value at build time. But this is likely not complex enough for your needs.

This probably warrants a feature request GitHub issue against PyOxidizer to discuss more options.

**Comment by gaborbernat** [comment]
@ofek you'll need to fix the merge conflicts :thinking: do you still plan to finish this? Thanks!

**Comment by gaborbernat** [comment]
Closing due to inactivity, we can reopen it when work picks up again on it.

**Pull request closed without merging**